specification/dartLangSpec.tex

```diff
@@ -2373,34 +2376,49 @@ \subsubsection{The Method \code{noSuchMethod}}
 \end{dartCode}
 
 \LMHash{}%
-Let $C$ be a concrete class and
-let $L$ be the library that contains the declaration of $C$.
-The member $m$ is \Index{noSuchMethod forwarded} in $C$ if{}f
+Let $C$ be a concrete class,
+let $L$ be the library that contains the declaration of $C$,
+and let $m$ be a name.
+Then $m$ is \Index{noSuchMethod forwarded} in $C$ if{}f
 one of the following is true:
 
 \begin{itemize}
-\item $C$ has a non-trivial \code{noSuchMethod},
-  the interface of $C$ contains $m$,
-  and $C$ has no concrete declaration of $m$
-  (\commentary{that is, no member $m$ is declared or inherited by $C$}).
+\item \textbf{Requested in program:}
+  $C$ has a non-trivial \code{noSuchMethod},
+  the interface of $C$ contains a member signature $S$ named $m$,
+  and $C$ has no concrete member named $m$ and accessible to $L$
+  that correctly overrides $S$
+  (\commentary{%
+    that is, no member named $m$ is declared or inherited by $C$,
+    or one is inherited, but it does not have the required signature%
+  }).
+  In this case we also say that $S$ is noSuchMethod forwarded.
 \item
-  % Inaccessible private methods are not present in the interface of a class,
-  % so we need to find a class that can access $m$.
+  \textbf{Forced by privacy:}
   There exists a direct or indirect superinterface
-  $D$ of $C$ which is declared in the library $L_2$,
-  the interface of $D$ contains $m$
-  (\commentary{which implies that $m$ is accessible to $L_2$}),
-  $m$ is inaccessible to $L$,
+  $D$ of $C$ which is declared in a library $L_2$ different from $L$,
+  the interface of $D$ contains a member signature $S$ named $m$,
+  $m$ is a private name,
   and no superclass of $C$ has
-  a concrete declaration of $m$ accessible to $L_2$.
+  a concrete member named $m$ accessible to $L_2$
+  that correctly overrides $S$.
+  In this case we also say that $S$ is noSuchMethod forwarded.
 \end{itemize}
 
 \LMHash{}%
 For a concrete class $C$, a
-\IndexCustom{\code{noSuchMethod} forwarder}{noSuchMethod forwarder}
-is implicitly induced for each member $m$
+\Index{noSuchMethod forwarder}
+is implicitly induced for each member signature
 which is noSuchMethod forwarded.
-This is a concrete member of $C$
+
+\LMHash{}%
+It is a compile-time error if the name $m$ is noSuchMethod forwarded
+in a concrete class $C$,
+and a superclass of $C$ has an accessible concrete declaration of $m$
+which is not a noSuchMethod forwarder.
+
+\LMHash{}%
+A noSuchMethod forwarder is a concrete member of $C$
 with the signature taken from the interface of $C$ respectively $D$ above,
```

Comment:
Do we need to say "respectively $D$" because the signature is also in the interface of $C$.
Also, since there might be more than on $D$ satisfying the requirement above for $m$ being privacy-nSM forwarded, we cannot speak about a particular $D$ here.

Reply:
Changed to avoid mentioning $D$. Added TODO comment saying that we should change the approach to privacy in general such that we are closer to name-mangling, such that we can actually say precisely what we want to say here.